### DISCUSSION TITLE
[Desglose] Tanque

### DISCUSSION THREAD

**Discussion by emrodk** [Godot-en-Espanol/Node-City#11] [Tareas]
Segundo ejercicio de desglose de tareas! Si no viste el primero (#9) te recomiendo que le pegues una leída donde analizamos el disparo. 

# Tanque (En proceso)
> Un tanque es un objeto que se desplaza a velocidad constante con un movimiento en grilla y colisiona. Puede disparar y soporta el uso de power ups para modificar su atributos

### Colisiona con
- Disparos. Recibe daño. Si su vida llega a 0 se destruye
- Bloques. Agua, Ladrillo y Metal
- Tanques. Solo bloquean el paso
- Power Ups (Triggers)

### Atributos identificados
- Vida
- Velocidad porque se mueve
- Dirección de movimiento. Se mueve en 4 direcciones.
- Cadencia de disparo o cada cuanto tiempo puede volver a disparar
- Potencia o velocidad de disparo

### Señales
- Cuando es destruido. Se debe contabilizar el score por lo que sería necesario que la señal contenga información del tanque
- Cuando detecta un tanque en rango de disparo (Player)

### Comportamientos
- Parpadea cuando pierde vida. Lo más simple
- Para mantenerlo simple, el tanque cuando es destruido desaparece.
- Cuando moverse o disparar puede ser controlado por el player o por una AI dependiendo del tipo del tanque
- Si el tanque es controlado por AI no interactua con power ups
- Deben ser capaz de detectar un enemigo a la distancia

### Tipos de tanque enemigos
- Básico: Su velocidad es baja, su potencia de disparo es baja. Se mueve haciendo un drunk walk y dispara constantemente
- Básico II : Misma AI que su versión anterior pero su velocidad y potencia de disparos es media.
- Rápido: Su velocidad es alta, su potencia de disparo es media. Se mueve en dirección a la base ignorando todo lo demás a fin de destruirla
- Ataque: Su velocidad es media y su potencia de dispara es alta. Se mueve en dirección a la base y si el player se encuentra cerca intenta destruirlo, dispara apenas visualiza un objetivo. Puede destruir bloques y acortar camino.
- Defensa: Su velocidad es media y su potencia de disparo es alta. Persigue al player, dispara apenas visualiza un objetivo
- Pesado: Su velocidad es lenta y su potencia de disparo es alta, su tanque está revestido de armadura lo que lo hece 4 veces más dificil de destruir. Puede romper bloques de metal. Se abre paso hasta la base y dispara al player apenas lo ve.

### Relación con otros elementos de juego
- Disparo. Lo instancia. Si esto conlleva un problema de performance se hará un pool de objetos

**Comment by gfazzito** [comment]
- Colisión con otros tanques
- interacción con power ups?
- indicación de daño: parpadea, se le indica una barra de vida, se le desprenden partes?
- al disparar en movimiento, pierde velocidad?

### Destrucción de tanques enemigos
- explota y desaparece, quedan los restos?
- Los restos pueden convertirse en un obstáculo en el camino? (quizas el tanque pesado unicamente)
- dejan power ups para recoger mediante un random?
- Los tanques grandes al morir, rompen un poco las paredes?

**Reply by emrodk** [reply]
Me gusta la idea de que exploten y rompan cosas, y que los restos molesten pero por las dudas lo dejamos como flores cosa de mantenerlo simple. Nota mental: armar lista de flores (cosas que agregan game juice o cosas "nice to have" es decir que suman pero no son estrictamente necesarias)

**Comment by joaquin-montesdeoca** [comment]
> **Señales**
> - Cuando es destruido. Se debe contabilizar el score

Una tontería que podemos hacer: por ahora solo se crea una señal cuando es destruido. Cuando esto pasa, podemos indicar el tipo de tanque que es, para que algún controlador se encargue de recibirlo:
- Jugador 1
- Jugador 2
- Tanque Enemigo Liviano
- Tanque Enemigo Mediano
- Tanto Enemigo Pesado

De esa manera, si es un Jugador, se encarga de ver si le quedan vidas y puede seguir jugando y si es un enemigo, se encarga de ingresar los puntos (siempre y cuando NO haya muerto por el power up de la granada) y de fijarse si quedan tanques para decidir si terminar el nivel o no.

> **Relación con otros elementos de juego**
> - Disparo. Lo instancia

No sé quién había sugerido esto, pero los disparos, se pueden guardar en un pool de objetos, para no tener que instanciarlos todo el tiempo.

**Tanque pesado**

Los tanques pesados podrían tener varios estados según la cantidad de energía que tienen, para simbolizar mejor que están perdiendo vida.

En el juego original, simplemente se limitaba a cambiar la paleta de colores. Quizás podríamos hacer eso (crear un tanque pesado estándar de color blanco a gris, que le cambiemos el modulate según la vida que le quede) o bien, crear distintos sprites para cada estado del tanque pesado.

**Objeto Comportamiento**

Esto es una sugerencia de como acomodar el objeto, pero podríamos tener un objeto comportamiento, a modo de Composite, que se instancia dentro del tanque y controla como se va a mover y atacar.

O sea, controla cómo se va a mover el objeto padre, el tanque:

- **Objeto Comportamiento Jugador**: simplemente se limita a hacerle caso al input de teclado o joystick que el jugador le dé.
- **Objeto Comportamiento Tanque Clásico**: el más simple de todos. Hace el Drunk Walk y dispara cada cierto tiempo.
- **Objeto Comportamiento Tanque Rápido**: creo que este era el más agresivo, porque se movía rápido y trataba de acercarse a la base del jugador.

Y esto varía según el tipo de tanque.

**Reply by emrodk** [reply]
Creo haber agregado la mayoría. Sobre el pool es evitable a menos que perdamos fps, lo del objeto comportamiento es una buena idea y creo que es bastante escalable.  Fijate los tipos de tanque que te parecen, los puse siguiendo la idea de AI diferentes

**Reply by emrodk** [reply]
Respecto al tanque pesado, creo que lo mejor es jugar con la paleta. SI fuera por mí se hacen todos los estados pero mejor no sobrecargar a gabo jajaj

**Reply by joaquin-montesdeoca** [reply]
Una boludez, cambié la frase inicial que escribí porque al releerlo me sonaba como que decía que lo de las señales era "una tontería" y me sonó un poco agresivo.

Simplemente, me refería a que era una "tontería" o "simpleza" lo que yo estaba a punto de sugerir.

Nada, eso.

**Reply by emrodk** [reply]
jaja tranqui que se entendió pero celebro la honestidad :heart:
